anomalib/utils/metrics/pro.py

```diff
@@ -0,0 +1,132 @@
+        """Compute the PRO score for the current batch."""
+
+        self.target.append(targets)
+        self.preds.append(predictions)
+
+    def compute(self) -> Tensor:
+        """Compute the macro average of the PRO score across all regions in all batches."""
+        target = dim_zero_cat(self.target)
+        preds = dim_zero_cat(self.preds)
+
+        if target.is_cuda:
+            comps = connected_components_gpu(target.unsqueeze(1))
+        else:
+            comps = connected_components_cpu(target.unsqueeze(1))
+
+        pro = pro_score(preds, comps, threshold=self.threshold)
+        return pro
+
+
+def pro_score(predictions: Tensor, comps: Tensor, threshold: float = 0.5) -> Tensor:
+    """Calculate the PRO score for a batch of predictions.
+
+    Args:
+        predictions (Tensor): Predicted anomaly masks (Bx1xHxW)
+        comps: (Tensor): Labeled connected components (BxHxW). The components should be labeled from 0 to N
+        threshold (float): When predictions are passed as float, the threshold is used to binarize the predictions.
+
+    Returns:
+        Tensor: Scalar value representing the average PRO score for the input batch.
+    """
+    if predictions.dtype == torch.float:
+        predictions = predictions > threshold
+
+    n_comps = len(comps.unique())
+
+    preds = comps.clone()
+    preds[~predictions] = 0
+    if n_comps == 1:  # only background
+        return torch.Tensor([1.0])
+    pro = recall(preds.flatten(), comps.flatten(), num_classes=n_comps, average="macro", ignore_index=0)
+    return pro
+
+
+def connected_components_gpu(binary_input: torch.Tensor, max_iterations: int = 1000) -> Tensor:
+    """Pytorch implementation for Connected Component Labeling on GPU.
+
+    Args:
+        binary_input (Tensor): Binary input data from which we want to extract connected components (Bx1xHxW)
+        max_iterations (int): Maximum number of iterations used in the connected component computaion.
+
+    Returns:
+        Tensor: Components labeled from 0 to N.
+    """
+    mask = binary_input.bool()
+
+    batch, _, height, width = binary_input.shape
+    components = torch.arange(batch * height * width, device=binary_input.device, dtype=torch.float).reshape(
+        (batch, 1, height, width)
+    )
+    components[~mask] = 0
+
+    converged = False
+    for _ in range(max_iterations):
+        previous = components.clone()
+        components[mask] = F.max_pool2d(components, kernel_size=3, stride=1, padding=1)[mask]
+        if torch.all(torch.eq(components, previous)):
+            converged = True
+            break
+
+    if not converged:
+        warnings.warn(
+            f"Max iterations ({max_iterations}) reached before converging. Connected component results may be "
+            f"inaccurate. Consider increasing the maximum number of iterations."
+        )
+
+    # remap component values from 0 to N
+    labels = components.unique()
+    for new_label, old_label in enumerate(labels):
```

Comment:
Not sure if I quite understood this correctly. When computing connected components in cpu, we go over each mask in the batch and find all the connected components. Then each CC is given a unique label across the image across each batch. I may be wrong here but won't maxpool give similar values across images in the batch. In that case the number of unique component labels might be lower than actual components. And what's the need to relabel the components as it is done here?
Also, why can't we use kornia's CC?

Reply:
The reason to re-label the components is that the `pro_score` function assumes that the components passed as the `comps` argument are labeled from 0 to N, where 0 is background and N is the total number of connected components in the batch. Because the connected component labeling algorithms that we use do not produce label maps that satisfy this requirement, we need to re-label them before passing them to the pro score computation. Since the cc labeling algorithms on cpu and gpu produce label maps with different characteristics, this re-labeling method also differs.

On cpu, we apply the connected component labeling on the image-level and then stack the labeled components back together in the batch dimension. So every connected component within the image will have a unique label, but labels will be duplicated across the batch. We need to re-label so that the labels are not only unique across the image, but also across the full batch.

On GPU, the connected component labeling works by assigning a unique label id to each pixel and then eliminating ids by merging connected regions together through iterative max pooling. This algorithm is applied on the full batch, so after the labeling, every connected component in the complete batch will have a unique label. So the number of unique values in the label map is equal to the number of connected components. However, due to the initial assignment of component id's [here](https://github.com/openvinotoolkit/anomalib/blob/2a29e535332fdd2e79a0546373714382f87a7b24/anomalib/utils/metrics/pro.py#L87), the components will not be labeled from 0 to N, but rather b. So we still need to re-label, but this time we can just create a mapping from the original ids to the [0, N] range.

Reply:
I just had a look at the Kornia implementation and it's essentially the same algorithm, so we could use the Kornia one instead (we'll still need to perform the re-labeling though).

Reply:
Thanks for the detailed explanation. It is clear now 🙂